packages/augur-core/src/contracts/para/AMMExchange.sol

```diff
@@ -8,236 +8,251 @@ import 'ROOT/ICash.sol';
-			_amounts[0] = _invalidAmount != 0 ? _invalidAmount : _noAmount;
-			_amounts[1] = _invalidAmount != 0 ? _yesAmount : _noAmount;
-		} else {
-			_tokenIds[0] = INVALID;
-			_tokenIds[1] = NO;
-			_tokenIds[2] = YES;
-			_amounts[0] = _invalidAmount;
-			_amounts[1] = _noAmount;
-			_amounts[2] = _yesAmount;
-		}
-		shareToken.unsafeBatchTransferFrom(_from, _to, _tokenIds, _amounts);
-	}
-
-	// babylonian method (https://en.wikipedia.org/wiki/Methods_of_computing_square_roots#Babylonian_method)
-	function sqrt(uint y) private pure returns (uint z) {
-		if (y > 3) {
-			z = y;
-			uint x = y / 2 + 1;
-			while (x < z) {
-				z = x;
-				x = (y / x + x) / 2;
-			}
-		} else if (y != 0) {
-			z = 1;
-		}
-	}
-
-	// TODO Switch sqrt to the constant time algo below
-	// Return value in range [0, 0x10000].
-	/*
-	uint32_t o1_32bit_sqrt_ver3(const uint32_t x) {
-		uint32_t s = 0, b = 1u << 15;
-		
-		// Compute the rounded-down square root:
-		while (b) {
-			const uint32_t t = (s + b); 
-			if (t * t <= x)  s = t;
-			b >>= 1;
-		}
-		return s;
-	}
-	*/
-
-	// TODO implement fee. Likely just making this adjustment everywhere: poolConstant = poolYes.mul(poolNo).mul(1000).div(997)
-	// TODO add method to exit a position based on shares
-	// TODO add convenience view methods to see what current rates are for swapping / entering a position
-
-	function onTokenTransfer(address _from, address _to, uint256 _value) internal {}
-}
\ No newline at end of file
+    using SafeMathUint256 for uint256;
+
+    ICash public cash;
+    IParaShareToken public shareToken;
+    IMarket public augurMarket;
+    uint256 public numTicks;
+    uint256 public INVALID;
+    uint256 public NO;
+    uint256 public YES;
+    uint256 public fee; // [0-1000] how many thousandths of swaps should be kept as fees
+    uint256 private liquidityConstant; // defaults to zero
+
+    constructor(IMarket _market, IParaShareToken _shareToken, uint256 _fee) public {
+        cash = _shareToken.cash();
+        shareToken = _shareToken;
+        augurMarket = _market;
+        numTicks = _market.getNumTicks();
+        cash.approve(address(_shareToken), 2**256-1);
+        cash.approve(address(_shareToken.augur()), 2**256-1);
+        INVALID = _shareToken.getTokenId(_market, 0);
+        NO = _shareToken.getTokenId(_market, 1);
+        YES = _shareToken.getTokenId(_market, 2);
+        fee = _fee;
+    }
+
+    // Adds shares to the liquidity pool by minting complete sets.
+    function addLiquidity(uint256 _sharesToBuy) external {
+        cash.transferFrom(msg.sender, address(this), _sharesToBuy.mul(numTicks));
+        shareToken.publicBuyCompleteSets(augurMarket, _sharesToBuy);
+
+        uint256 _newLiquidityConstant = calculateLiquidityConstant();
+
+        if (liquidityConstant == 0) {
+            // The formula for issuing LP tokens breaks down when the supply (and therefore liquidityConstant) are zero.
+            _mint(msg.sender, _newLiquidityConstant);
```

Comment:
There is a known bug in Uniswap v1 where someone can permanently break an exchange by setting it up with a tiny bit of liquidity.  This may not matter for this project like it does with Uniswap, especially if there can be multiple exchanges per market.

Reply:
I think it will still matter. I assume it's this? https://github.com/Uniswap/uniswap-v2-core/blob/master/contracts/UniswapV2Pair.sol#L121

Reply:
Yeah, that is the one.  It doesn't matter if you can have multiple exchanges per pair, since if someone creates a broken pair you can just create another one to use instead.